src/data/real/pi/wallis.lean

```diff
@@ -5,70 +5,228 @@ Authors: Hanting Zhang
-    { rw mul_one } },
-  convert integral_sin_pow_div_tendsto_one,
-  funext,
   rw [integral_sin_pow_even, integral_sin_pow_odd, mul_div_mul_comm, ←prod_div_distrib, inv_div],
-  congr' with i,
-  rw [div_div_div_comm, div_div_eq_mul_div, mul_div_assoc],
+  simp_rw [div_div_div_comm, div_div_eq_mul_div, mul_div_assoc],
+  refl,
+end
+
+lemma W_le (k : ℕ) : W k ≤ π / 2 :=
+begin
+  rw [←div_le_one pi_div_two_pos, div_eq_inv_mul],
+  rw [W_eq_integral_sin_pow_div_integral_sin_pow, div_le_one (integral_sin_pow_pos _)],
+  apply integral_sin_pow_succ_le,
+end
+
+lemma W_ge (k : ℕ) : ((2:ℝ) * k + 1) / (2 * k + 2) * (π / 2) ≤ W k :=
+begin
+  rw [←le_div_iff pi_div_two_pos, div_eq_inv_mul (W k) _],
+  rw [W_eq_integral_sin_pow_div_integral_sin_pow, le_div_iff (integral_sin_pow_pos _)],
+  convert integral_sin_pow_succ_le (2 * k + 1),
+  rw [integral_sin_pow (2 * k)],
+  simp only [sin_zero, zero_pow', ne.def, nat.succ_ne_zero, not_false_iff, zero_mul, sin_pi,
+    tsub_zero, nat.cast_mul, nat.cast_bit0, algebra_map.coe_one, zero_div, zero_add],
+end
+
+lemma tendsto_W_nhds_pi_div_two : tendsto W at_top (𝓝 $ π / 2) :=
+begin
+  refine tendsto_of_tendsto_of_tendsto_of_le_of_le _ tendsto_const_nhds W_ge W_le,
+  have : 𝓝 (π / 2) = 𝓝 ((1 - 0) * (π / 2)) := by rw [sub_zero, one_mul], rw this,
+  refine tendsto.mul _ tendsto_const_nhds,
+  have h : ∀ (n:ℕ), ((2:ℝ) * n + 1) / (2 * n + 2) = 1 - 1 / (2 * n + 2),
+  { intro n,
+    rw [sub_div' _ _ _ (ne_of_gt (add_pos_of_nonneg_of_pos
+      (mul_nonneg ((two_pos : 0 < (2:ℝ)).le) (nat.cast_nonneg _)) two_pos)), one_mul],
+    congr' 1, ring },
+  simp_rw h,
+  refine (tendsto_const_nhds.div_at_top _).const_sub _,
+  refine tendsto.at_top_add _ tendsto_const_nhds,
+  exact tendsto_coe_nat_at_top_at_top.const_mul_at_top two_pos,
+end
+
+lemma W_eq_mul_sq (k : ℕ) :
+  W k = (2 * k + 1) * (∏ i in range k, ((2:ℝ) * i + 2) / (2 * i + 3)) ^ 2 :=
+begin
+  induction k with k hk,
+  { simp [W], },
+  { unfold W at *,
+    rw [prod_range_succ, prod_range_succ, hk],
+    suffices : ∀ (α : ℝ),
+      (2 * ↑k + 1) * α ^ 2 * ((2 * ↑k + 2) / (2 * ↑k + 1) * ((2 * ↑k + 2) / (2 * ↑k + 3))) =
+      (2 * ↑(k.succ) + 1) * (α * ((2 * ↑k + 2) / (2 * ↑k + 3))) ^ 2,
+    { rw this },
+    intro α,
+    have a : (2 * ↑k + 1 : ℝ) ≠ 0 := by positivity,
+    have b : (2 * ↑k + 3 : ℝ) ≠ 0 := by positivity,
+    field_simp, ring }
 end
 
+lemma integral_sin_pow_odd_sq_eq (k : ℕ) :
+  (∫ x in 0..π, sin x ^ (2 * k + 1)) ^ 2 = 4 * W k / (2 * k + 1) :=
+begin
+  rw integral_sin_pow_odd,
+  have B := W_eq_mul_sq k,
+  rw [mul_comm (2 * (k:ℝ) + 1) _, ←div_eq_iff] at B,
+  rw [mul_pow, ←B],
+  ring,
+  linarith [(nat.cast_nonneg k : 0 ≤ (k:ℝ))],
+end
+
+lemma integral_sin_pow_even_sq_eq (k : ℕ) :
+  (∫ x in 0..π, sin x ^ (2 * k)) ^ 2 = π ^ 2 / (2 * k + 1) / W k :=
+begin
+  induction k with k hk,
+  { dsimp only [W],
+    simp },
+  { have np : 0 < 2 * (k:ℝ) + 1 := by linarith [(nat.cast_nonneg k : 0 ≤ (k:ℝ))],
```

Comment:
```suggestion
  { have np : 0 < 2 * (k:ℝ) + 1, by linarith [(nat.cast_nonneg k : 0 ≤ (k:ℝ))],
```
no need to go out and then back inside tactic mode. Same thing many times. And in fact you should probably use `positivity` here (and also in `np2` and in `np3`), as it is much more lightweight than `linarith`.